src/Expander.hs

```diff
@@ -456,34 +459,36 @@ initializeKernel = do
               Just _ ->
                 error "TODO throw real error - already expanding a module"
               Nothing -> do
-                bodyPtr <- liftIO newModBodyPtr
+                bodyPtr <- liftIO newDeclTreePtr
                 modifyState $ set expanderModuleTop (Just bodyPtr)
                 Stx _ _ (_ :: Syntax, name, imports, body, exports) <- mustHaveEntries stx
                 _actualName <- mustBeIdent name
 
                 resolveImports imports
 
-                forkExpandDecls bodyPtr body
+                vp <- newDeclValidityPtr
+                expandDeclForms bodyPtr mempty vp body
 
                 buildExports exports
                 pure ()
         )
       ]
 
-    declPrims :: [(Text, Scope -> DeclPtr -> DeclValidityPtr -> Syntax -> Expand ())]
+    declPrims :: [(Text, DeclTreePtr -> ScopeSet -> DeclValidityPtr -> Syntax -> Expand ())]
     declPrims =
       [ ( "define"
-        , \ sc dest pdest stx -> do
+        , \ dest scs vp stx -> do
             p <- currentPhase
             Stx _ _ (_, varStx, expr) <- mustHaveEntries stx
+            sc <- freshScope $ T.pack $ "For definition at " ++ shortShow (stxLoc stx)
```

Comment:
Instead of receiving a `Scope` which can be used to bind a single new identifier, each `declPrim` now creates as many `Scope`s as they need, and the fills the `DeclValidityPtr` with a `ScopeSet` containing all those `Scope`s (with the proper phase for each). This will make it possible for `no-decls` to create 0 and for `decls` to create many.

Since each `Scope` created in that tree of declarations needs to be visible from all the later scopes, we need to accumulate more an more `Scope`s as we traverse that tree. To do that, each `declPrim` is now receiving an input `ScopeSet` as well, consisting of all the `Scope`s which have been created so far, and then it adds its own `Scope`s to this input `ScopeSet` and puts the resulting output `ScopeSet` in the `DeclValidityPtr`. I think I might be able to make this simpler.

Reply:
I did manage to make this simpler: instead of receiving a `ScopeSet` and adding their own scopes to it, most declaration forms do not receive a `ScopeSet`, and so they only need to fill their `DeclValidityPtr` with the scopes they introduce themselves. The exception to that is the `ExpandFormDecls` task, which now receives both a `ScopeSet` _and_ an input `DeclValidityPtr` _and_ an output `DeclValidityPtr`. The comments I wrote there and on `expandDeclForm[s]` are hopefully clear enough about how this works.

Reply:
They are indeed clear enough. Thanks!